### PR TITLE
Implement MatmulInteger on GPU

### PR DESCRIPTION
**Pull request by yufenglee**
**Description**: Describe your changes.

**Motivation and Context**
- Why is this change required? What problem does it solve?
- If it fixes an open issue, please link to the issue here.

### CODE REVIEW

File: onnxruntime/core/providers/cuda/math/matmul_integer.cc [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+  const int8_t* b_ptr = b->template Data<int8_t>();
+  int32_t* output_ptr = Y->template MutableData<int32_t>();
+
+  // validate zero points
+  const int8_t* a_offset = nullptr;
+  const int8_t* b_offset = nullptr;
+  if (has_a_zero_point_) {
+    auto a_zero_point = ctx->Input<Tensor>(2);
+    ORT_ENFORCE(IsScalarOr1ElementVector(a_zero_point),
+                "MatmulInteger : input1 zero point must be a scalar or 1D tensor of size 1");
```
Comment (skottmckay) [inline onnxruntime/core/providers/cuda/math/matmul_integer.cc]: The spec says 
a_zero_point (optional) : T1
Zero point tensor for input 'A'. It's optional and default value is 0. It could be a scalar or a 1-D tensor, which means a per-tensor or per-row quantization. If it's a 1-D tensor, its number of elements should be equal to the number of rows of input 'A'.

This check is that there's only one element isn't it, so if the number of rows in input 'A' was > 1 shouldn't the number of elements here match that count?

File: onnxruntime/test/providers/cpu/math/matmul_integer_test.cc [hunk 2]
```diff
@@ -32,6 +33,203 @@ TEST(MatmulIntegerOpTest, MatMulInteger) {
   test.AddOutput<int32_t>("T3", {1, 1}, {-1});
   test.Run();
 }
+
+// clang-format off
+TEST(MatmulIntegerOpTest, MatMulInteger_int8_t) {
```
Comment (skottmckay) [inline onnxruntime/test/providers/cpu/math/matmul_integer_test.cc]: Why is this necessary?

File: onnxruntime/test/providers/cpu/math/matmul_integer_test.cc [hunk 3]
```diff
@@ -32,6 +33,203 @@ TEST(MatmulIntegerOpTest, MatMulInteger) {
...
+  OpTester test("MatMulInteger", 10);
```
Comment (skottmckay) [inline onnxruntime/test/providers/cpu/math/matmul_integer_test.cc]: Do any of the tests validate this?
> The production MUST never overflow. The accumulation may overflow if and only if in 32 bits.

File: onnxruntime/core/providers/cuda/math/matmul_integer.cc [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+
+  for (int batch = 0; batch < helper.OutputOffsets().size(); batch++) {
+    CUBLAS_RETURN_IF_ERROR(cublasGemmEx(
+        Base::CublasHandle(),
+        CUBLAS_OP_N,
+        CUBLAS_OP_N,
+        static_cast<int>(helper.N()),
+        static_cast<int>(helper.M()),
+        static_cast<int>(helper.K()),
+        &alpha,
```
Comment (skottmckay) [inline onnxruntime/core/providers/cuda/math/matmul_integer.cc]: Any chance this conversion from int64_t to int would ever overflow and should be checked?

File: onnxruntime/test/providers/cpu/math/matmul_integer_test.cc [hunk 5]
```diff
@@ -32,6 +33,203 @@ TEST(MatmulIntegerOpTest, MatMulInteger) {
...
+  test.AddInput<int8_t>("a_zero_point", {}, {5});
+  test.AddInput<int8_t>("b_zero_point", {}, {5});
+  test.AddOutput<int32_t>("T3",
+                          {2, 4},
+                          { -55,  16, 89, -44,
+                            122, 154, 68, -39});
+
+  std::vector<std::unique_ptr<IExecutionProvider>> execution_providers;
+  execution_providers.push_back( DefaultCudaExecutionProvider() );
+  test.Run( OpTester::ExpectResult::kExpectSuccess, "", {}, nullptr, &execution_providers );
```
Comment (skottmckay) [inline onnxruntime/test/providers/cpu/math/matmul_integer_test.cc]: Usually we run all tests with the CPU EP as well. Should it be added here as well to validate that output from the CPU implementation matches the CUDA implementation?

File: onnxruntime/core/providers/cuda/math/matmul_integer.cc [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+    ORT_ENFORCE(IsScalarOr1ElementVector(a_zero_point),
+                "MatmulInteger : input1 zero point must be a scalar or 1D tensor of size 1");
+    a_offset = a_zero_point->template Data<int8_t>();
+  }
+  if (has_b_zero_point_) {
+    auto b_zero_point = ctx->Input<Tensor>(3);
+    ORT_ENFORCE(IsScalarOr1ElementVector(b_zero_point),
+                "MatmulInteger : input2 zero point must be a scalar or 1D tensor of size 1");
+    b_offset = b_zero_point->template Data<int8_t>();
+  }
```
Comment (skottmckay) [inline onnxruntime/core/providers/cuda/math/matmul_integer.cc]: Slightly confused by the spec as it describes a_zero_point as a 'zero point tensor' and b_zero_point as a 'scale tensor'. 

>b_zero_point (optional) : T2
>Scale tensor for input 'B'. 

Are they both used as a zero point tensors here?

File: onnxruntime/core/providers/cuda/math/matmul_integer.cc (hunk 1 above)
Comment (yufenglee) [reply onnxruntime/core/providers/cuda/math/matmul_integer.cc]: Yes, that's right. However, the QuantizeLinear [https://github.com/onnx/onnx/blob/master/docs/Operators.md#quantizelinear](url) doesn't support per-axis quantization in ONNX, which means there won't be a per-row/per-column MatMulInteger in practice. That's why our CPU also implements as this. 
We need to fix the QuantizeLinear and this implementation as well.

File: onnxruntime/test/providers/cpu/math/matmul_integer_test.cc (hunk 3 above)
Comment (yufenglee) [reply onnxruntime/test/providers/cpu/math/matmul_integer_test.cc]: It is guaranteed by our compute type. One int8 * int8 won't overflow with compute type int32.

File: onnxruntime/core/providers/cuda/math/matmul_integer.cc (hunk 6 above)
Comment (yufenglee) [reply onnxruntime/core/providers/cuda/math/matmul_integer.cc]: > Slightly confused by the spec as it describes a_zero_point as a 'zero point tensor' and b_zero_point as a 'scale tensor'.
> 
> > b_zero_point (optional) : T2
> > Scale tensor for input 'B'.
> 
> Are they both used as a zero point tensors here?

a_zero_point and b_zero_point is the zero point for Matrix A and B respectively, and a_scale and b_scale is the scale for A and B.

File: onnxruntime/core/providers/cuda/math/matmul_integer.cc (hunk 4 above)
Comment (yufenglee) [reply onnxruntime/core/providers/cuda/math/matmul_integer.cc]: a matrix with a dimension > 2^31 is extremely rare in practice.

File: onnxruntime/test/providers/cpu/math/matmul_integer_test.cc (hunk 5 above)
Comment (skottmckay) [reply onnxruntime/test/providers/cpu/math/matmul_integer_test.cc]: Unsupported on CPU.
